File: util/debug_test.go
package util

import (
	"io"
	"os"
	"testing"
)

func captureStderr(t *testing.T) func() string {
	t.Helper()

	orig := os.Stderr
	r, w, err := os.Pipe()
	if err != nil {
		t.Fatalf("os.Pipe failed: %v", err)
	}
	os.Stderr = w

	return func() string {
		t.Helper()
		_ = w.Close()
		os.Stderr = orig
		b, err := io.ReadAll(r)
		if err != nil {
			t.Fatalf("ReadAll failed: %v", err)
		}
		_ = r.Close()
		return string(b)
	}
}

func TestNewDebugDefaultAndNamespaceOutput(t *testing.T) {
	d := newDebugNotInitialized("test:ns")
	done := captureStderr(t)

	d.Print("before")
	applyDebug("golar:test:ns", []*Debug{d})
	d.Print("after")

	got := done()
	want := "golar:test:ns after\n"
	if got != want {
		t.Fatalf("unexpected stderr output:\nwant %q\ngot  %q", want, got)
	}
}

func TestApplyDebugMatchesAndSkips(t *testing.T) {
	matched := newDebugNotInitialized("api:users")
	skipped := newDebugNotInitialized("api:internal")
	unmatched := newDebugNotInitialized("worker:queue")
	done := captureStderr(t)

	applyDebug("golar:api:*,-golar:api:internal", []*Debug{matched, skipped, unmatched})
	matched.Print("matched")
	skipped.Print("skipped")
	unmatched.Print("unmatched")

	got := done()
	want := "golar:api:users matched\n"
	if got != want {
		t.Fatalf("unexpected stderr output:\nwant %q\ngot  %q", want, got)
	}
}

func TestApplyDebugLeadingSkipThenMatch(t *testing.T) {
	exact := newDebugNotInitialized("api")
	child := newDebugNotInitialized("api:users")
	done := captureStderr(t)

	applyDebug("-golar:api,golar:api*", []*Debug{exact, child})
	exact.Print("exact")
	child.Print("child")

	got := done()
	want := "golar:api:users child\n"
	if got != want {
		t.Fatalf("unexpected stderr output:\nwant %q\ngot  %q", want, got)
	}
}

func TestApplyDebugEmptyStringNoOp(t *testing.T) {
	api := newDebugNotInitialized("api")
	worker := newDebugNotInitialized("worker")
	done := captureStderr(t)

	applyDebug("golar:api", []*Debug{api, worker})
	applyDebug("", []*Debug{api, worker})
	api.Print("api")
	worker.Print("worker")

	got := done()
	want := "golar:api api\n"
	if got != want {
		t.Fatalf("unexpected stderr output:\nwant %q\ngot  %q", want, got)
	}
}

func TestApplyDebugSkipsEmptyParts(t *testing.T) {
	api := newDebugNotInitialized("api")
	worker := newDebugNotInitialized("worker")
	done := captureStderr(t)

	applyDebug(",,golar:api,,", []*Debug{api, worker})
	api.Print("api")
	worker.Print("worker")

	got := done()
	want := "golar:api api\n"
	if got != want {
		t.Fatalf("unexpected stderr output:\nwant %q\ngot  %q", want, got)
	}
}

func TestApplyDebugSkipsBareDashPart(t *testing.T) {
	api := newDebugNotInitialized("api")
	worker := newDebugNotInitialized("worker")
	done := captureStderr(t)

	applyDebug("-,golar:api", []*Debug{api, worker})
	api.Print("api")
	worker.Print("worker")

	got := done()
	want := "golar:api api\n"
	if got != want {
		t.Fatalf("unexpected stderr output:\nwant %q\ngot  %q", want, got)
	}
}

func TestApplyDebugAddsAlternationForMultipleParts(t *testing.T) {
	api := newDebugNotInitialized("api")
	worker := newDebugNotInitialized("worker")
	internal := newDebugNotInitialized("internal")
	other := newDebugNotInitialized("other")
	done := captureStderr(t)

	applyDebug("golar:api,golar:worker,-golar:internal,-golar:other", []*Debug{api, worker, internal, other})
	api.Print("api")
	worker.Print("worker")
	internal.Print("internal")
	other.Print("other")

	got := done()
	want := "golar:api api\ngolar:worker worker\n"
	if got != want {
		t.Fatalf("unexpected stderr output:\nwant %q\ngot  %q", want, got)
	}
}

func TestApplyDebugSkipOnlyReturnsWithoutChanges(t *testing.T) {
	api := newDebugNotInitialized("api")
	worker := newDebugNotInitialized("worker")
	done := captureStderr(t)

	applyDebug("golar:api", []*Debug{api, worker})
	applyDebug("-golar:api,-golar:worker", []*Debug{api, worker})
	api.Print("api")
	worker.Print("worker")

	got := done()
	want := "golar:api api\n"
	if got != want {
		t.Fatalf("unexpected stderr output:\nwant %q\ngot  %q", want, got)
	}
}

func TestApplyDebugWildcardMatchWithWildcardSkip(t *testing.T) {
	web := newDebugNotInitialized("service:web")
	dbRead := newDebugNotInitialized("service:db:read")
	dbWrite := newDebugNotInitialized("service:db:write")
	done := captureStderr(t)

	applyDebug("golar:service:*,-golar:service:db*", []*Debug{web, dbRead, dbWrite})
	web.Print("web")
	dbRead.Print("db-read")
	dbWrite.Print("db-write")

	got := done()
	want := "golar:service:web web\n"
	if got != want {
		t.Fatalf("unexpected stderr output:\nwant %q\ngot  %q", want, got)
	}
}

func TestPrintAndPrintf(t *testing.T) {
	done := captureStderr(t)
	d := newDebugNotInitialized("test:ns")

	applyDebug("golar:test:ns", []*Debug{d})
	d.Print("hello", "world")
	d.Printf("value=%d", 7)

	got := done()
	want := "golar:test:ns hello world\ngolar:test:ns value=7\n"
	if got != want {
		t.Fatalf("unexpected stderr output:\nwant %q\ngot  %q", want, got)
	}
}

func TestPrintAndPrintfWhenNotMatched(t *testing.T) {
	done := captureStderr(t)
	d := newDebugNotInitialized("test:disabled")

	applyDebug("golar:other", []*Debug{d})
	d.Print("hello")
	d.Printf("value=%d", 7)

	if got := done(); got != "" {
		t.Fatalf("expected no output for unmatched debug, got %q", got)
	}
}

File: packages/util/src/debug.test.ts
import { describe, expect, it, vi } from 'vitest'

import { createDebugFromString } from './debug.ts'

function captureStderr(fn: () => void): string {
	let stderr = ''
	const spy = vi.spyOn(process.stderr, 'write').mockImplementation(((
		chunk: unknown,
	) => {
		stderr += String(chunk)
		return true
	}) as never)

	try {
		fn()
		return stderr
	} finally {
		spy.mockRestore()
	}
}

describe('logger', () => {
	it('defaults to disabled for empty debug', () => {
		const output = captureStderr(() => {
			const l = createDebugFromString('', 'test:ns')
			l.print('hello')
			l.printf('value=%d', 7)
		})
		expect(output).toBe('')
	})

	it('matches and skips', () => {
		const output = captureStderr(() => {
			createDebugFromString(
				'golar:api:*,-golar:api:internal',
				'api:users',
			).print('m')
			createDebugFromString(
				'golar:api:*,-golar:api:internal',
				'api:internal',
			).print('s')
			createDebugFromString(
				'golar:api:*,-golar:api:internal',
				'worker:queue',
			).print('u')
		})
		expect(output).toBe('golar:api:users m\ngolar:worker:queue u\n')
	})

	it('leading skip then match', () => {
		const output = captureStderr(() => {
			createDebugFromString('-golar:api,golar:api*', 'api').print('exact')
			createDebugFromString('-golar:api,golar:api*', 'api:users').print('child')
		})
		expect(output).toBe('golar:api:users child\n')
	})

	it('skips empty parts', () => {
		const output = captureStderr(() => {
			createDebugFromString(',,golar:api,,', 'api').print('a')
			createDebugFromString(',,golar:api,,', 'worker').print('w')
		})
		expect(output).toBe('golar:api a\ngolar:worker w\n')
	})

	it('skips bare dash part', () => {
		const output = captureStderr(() => {
			createDebugFromString('-,golar:api', 'api').print('a')
			createDebugFromString('-,golar:api', 'worker').print('w')
		})
		expect(output).toBe('golar:api a\ngolar:worker w\n')
	})

	it('adds alternation for multiple parts', () => {
		const output = captureStderr(() => {
			createDebugFromString(
				'golar:api,golar:worker,-golar:internal,-golar:other',
				'api',
			).print('a')
			createDebugFromString(
				'golar:api,golar:worker,-golar:internal,-golar:other',
				'worker',
			).print('w')
			createDebugFromString(
				'golar:api,golar:worker,-golar:internal,-golar:other',
				'internal',
			).print('i')
			createDebugFromString(
				'golar:api,golar:worker,-golar:internal,-golar:other',
				'other',
			).print('o')
		})
		expect(output).toBe('golar:api a\ngolar:worker w\n')
	})

	it('skip only leaves logger disabled', () => {
		const output = captureStderr(() => {
			createDebugFromString('-api,-worker', 'api').print('a')
			createDebugFromString('-api,-worker', 'worker').print('w')
		})
		expect(output).toBe('')
	})

	it('wildcard match with wildcard skip', () => {
		const output = captureStderr(() => {
			createDebugFromString(
				'golar:service:*,-golar:service:db*',
				'service:web',
			).print('web')
			createDebugFromString(
				'golar:service:*,-golar:service:db*',
				'service:db:read',
			).print('read')
			createDebugFromString(
				'golar:service:*,-golar:service:db*',
				'service:db:write',
			).print('write')
		})
		expect(output).toBe('golar:service:web web\n')
	})

	it('print and printf', () => {
		const output = captureStderr(() => {
			const l = createDebugFromString('test:*', 'test:ns')
			l.print('hello', 'world')
			l.printf('value=%d', 7)
		})
		expect(output).toBe('golar:test:ns hello world\ngolar:test:ns value=7\n')
	})

	it('print and printf when disabled', () => {
		const output = captureStderr(() => {
			const l = createDebugFromString('', 'test:ns')
			l.print('hello')
			l.printf('value=%d', 7)
		})
		expect(output).toBe('')
	})
})
